File: es17/es17.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	NONE = iota
	CLAY
	SAND
	WATER
	MaxInt = int(^uint(0) >> 1)
	MinInt = -MaxInt - 1
)

type Map struct {
	xmin, ymin    int
	xmax, ymax    int
	width, height int

	data [][]int
}

func (m *Map) minmax(x, y int) {
	if m.xmin > x {
		m.xmin = x
	}
	if m.xmax < x {
		m.xmax = x
	}
	if m.ymin > y {
		m.ymin = y
	}
	if m.ymax < y {
		m.ymax = y
	}
}

func (m *Map) Get(x, y int) int {
	if x >= m.xmin && x <= m.xmax && y >= m.ymin && y <= m.ymax {
		return m.data[y-m.ymin][x-m.xmin]
	}
	return NONE
}

func (m *Map) Set(x, y, v int) {
	if x >= m.xmin && x <= m.xmax && y >= m.ymin && y <= m.ymax {
		m.data[y-m.ymin][x-m.xmin] = v
	}
}

func (m *Map) Open(x, y int) bool {
	v := m.Get(x, y)
	return v == NONE || v == SAND
}

func (m *Map) Fill(x, y int) {
	if y > m.ymax || !m.Open(x, y) {
		return
	} else if !m.Open(x, y+1) {
		l := x
		for ; m.Open(l, y) && !m.Open(l, y+1); l-- {
			m.Set(l, y, SAND)
		}

		r := x + 1
		for ; m.Open(r, y) && !m.Open(r, y+1); r++ {
			m.Set(r, y, SAND)
		}

		if m.Open(l, y) || m.Open(r, y) {
			m.Fill(l, y)
			m.Fill(r, y)
		} else {
			l++
			for ; l < r; l++ {
				m.Set(l, y, WATER)
			}
			m.Fill(x, y-1)
		}
	} else if m.Get(x, y) == NONE {
		m.Set(x, y, SAND)
		m.Fill(x, y+1)
	}
}

func (m *Map) Count() (int, int) {
	sand := 0
	water := 0
	for y := 0; y < m.height; y++ {
		for x := 0; x < m.width; x++ {
			if m.data[y][x] == SAND {
				sand++
			} else if m.data[y][x] == WATER {
				water++
			}
		}
	}
	return sand, water
}

func (m *Map) Print() {
	for y := 0; y < m.height; y++ {
		for x := 0; x < m.width; x++ {
			switch m.data[y][x] {
			case NONE:
				fmt.Print(".")
			case CLAY:
				fmt.Print("#")
			case SAND:
				fmt.Print("|")
			case WATER:
				fmt.Print("~")
			}
		}
		fmt.Println()
	}
}

func LoadMap(input *os.File) Map {
	m := Map{xmin: MaxInt, ymin: MaxInt, xmax: MinInt, ymax: MinInt}
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()
		var a, b, c int
		_, err := fmt.Sscanf(line, "x=%d, y=%d..%d\n", &a, &b, &c)
		if err == nil {
			m.minmax(a, b)
			m.minmax(a, c)
		}

		_, err = fmt.Sscanf(line, "y=%d, x=%d..%d\n", &a, &b, &c)
		if err == nil {
			m.minmax(b, a)
			m.minmax(c, a)
		}
	}
	m.xmin--
	m.xmax++
	m.width = m.xmax - m.xmin + 1
	m.height = m.ymax - m.ymin + 1

	// build the map
	m.data = make([][]int, m.height)
	for y := 0; y < m.height; y++ {
		m.data[y] = make([]int, m.width)
	}

	input.Seek(0, 0)
	scanner = bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()

		var a, b, c int
		_, err := fmt.Sscanf(line, "x=%d, y=%d..%d\n", &a, &b, &c)
		if err == nil {
			for y := b; y <= c; y++ {
				m.Set(a, y, CLAY)
			}
		}

		_, err = fmt.Sscanf(line, "y=%d, x=%d..%d\n", &a, &b, &c)
		if err == nil {
			for x := b; x <= c; x++ {
				m.Set(x, a, CLAY)
			}
		}
	}
	return m
}

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <filename>\n", os.Args[0])
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	m := LoadMap(input)
	m.Fill(500, 0)
	// m.Print()
	sand, water := m.Count()
	fmt.Printf("sand %d, water %d, sum %d\n", sand, water, sand+water)

	return 0
}

File: es17/es17.py
#!/usr/bin/env python

import sys
import re

# constants
NONE = 0
CLAY = 1
SAND = 2
WATER = 3

class Map(object):
    def __init__(self):
        self.xmin = self.ymin = 0
        self.xmax = self.ymax = 0
        self.width = self.height = 0
        self.data = None

    def minmax(self, x, y):
        if self.xmin > x:
            self.xmin = x
        if self.xmax < x:
            self.xmax = x
        if self.ymin > y:
            self.ymin = y
        if self.ymax < y:
            self.ymax = y

    def set(self, x, y, value):
        if x >= self.xmin and x <= self.xmax and y >= self.ymin and y <= self.ymax:
            self.data[y - self.ymin][x - self.xmax] = value

    def get(self, x, y):
        if x >= self.xmin and x <= self.xmax and y >= self.ymin and y <= self.ymax:
            return self.data[y - self.ymin][x - self.xmax]
        return NONE

    def open(self, x, y):
        v = self.get(x, y)
        return v == NONE or v == SAND

    def fill(self, x, y):
        stack = [(x, y)]
        while len(stack) > 0:
            x, y = stack.pop()
            if y > self.ymax:
                pass
            elif not self.open(x, y):
                pass
            elif not self.open(x, y+1):
                l = x
                while self.open(l, y) and not self.open(l, y+1):
                    self.set(l, y, SAND)
                    l -= 1

                r = x + 1
                while self.open(r, y) and not self.open(r, y+1):
                    self.set(r, y, SAND)
                    r += 1

                if self.open(l, y) or self.open(r, y):
                    stack.append((l, y))
                    stack.append((r, y))
                else:
                    l += 1
                    while l < r:
                        self.set(l, y, WATER)
                        l += 1
                    stack.append((x, y-1))
            elif self.get(x, y) == NONE:
                self.set(x, y, SAND)
                stack.append((x, y+1))

    def count(self):
        sand = water = 0
        for y in range(self.height):
            for x in range(self.width):
                if self.data[y][x] == SAND:
                    sand += 1
                elif self.data[y][x] == WATER:
                    water += 1
        return sand, water

    def print(self):
        for y in range(self.height):
            l = []
            for x in range(self.width):
                t = self.data[y][x]
                if t == CLAY:
                    l.append("#")
                elif t == SAND:
                    l.append("|")
                elif t == WATER:
                    l.append("~")
                else:
                    l.append(".")
            print("".join(l))

    def load(self, file):
        vertical = re.compile(r"""x=(\d+), y=(\d+)..(\d+)""")
        horizontal = re.compile(r"""y=(\d+), x=(\d+)..(\d+)""")
        self.xmin = self.ymin = 10000
        self.xmax = self.ymax = -10000
        for line in file:
            m = vertical.match(line)
            if m:
                a, b, c = [int(x) for x in m.groups()]
                self.minmax(a, b)
                self.minmax(a, c)

            m = horizontal.match(line)
            if m:
                a, b, c = [int(x) for x in m.groups()]
                self.minmax(b, a)
                self.minmax(c, a)

        self.xmin -= 1
        self.xmax += 1
        self.width = self.xmax - self.xmin + 1
        self.height = self.ymax - self.ymin + 1
        self.data = [[0 for x in range(self.width)] for y in range(self.height)]

        file.seek(0, 0)
        for line in file:
            m = vertical.match(line)
            if m:
                a, b, c = [int(x) for x in m.groups()]
                for y in range(b, c+1):
                    self.set(a, y, CLAY)

            m = horizontal.match(line)
            if m:
                a, b, c = [int(x) for x in m.groups()]
                for x in range(b, c+1):
                    self.set(x, a, CLAY)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    m = Map()
    try:
        with open(sys.argv[1], "rt") as f:
            m.load(f)
    except:
        raise
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    m.fill(500,0)
    # m.print()
    sand, water = m.count()
    print("sand {}, water {}, sum {}".format(sand, water, sand + water))
